### PR TITLE
feat: Select all for synchronous select

### PR DESCRIPTION
**Pull request by cccs-RyanK**
### SUMMARY

This PR adds a 'select all' functionality to the synchronous Select component. This functionality has been requested by a large number of users and contributors. The select component has been previously split into sync and async components to simplify the code. The sync select component is the more widely used component, and a select all for the sync case is much easier to implement. This PR simply adds the option for selectAllEnabled only when the select component is in multi-select mode. The option is by default set to false. This is part of the greater task to rework the select component for the select all functionality outlined by @michael-s-molina in this PR: https://github.com/apache/superset/pull/20143

### TESTING INSTRUCTIONS

In storybook, go to the select component -> interactive select and set the mode to multi and selectAllEnabled to True

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache/superset/discussions/18247#discussioncomment-2467080
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/components/Select/utils.tsx
Comment (michael-s-molina) [inline superset-frontend/src/components/Select/utils.tsx:28]: For compatibility reasons, I think in this first phase we should send all values to `onChange` when selecting "Select all". It would be the same as our current behavior where a user can manually select all items. In the next phase, we should come up with a smarter way of representing the "Select All" value but this will require backend changes to interpret this value. My [original PR](https://github.com/apache/superset/pull/20143) has an item about this:

> Handle the payload size. We shouldn't send all selected items to the server-side because we don't know how many can exist, which could result in huge payload size. What we need to do is to come up with a payload that represents the user selection in an efficient way. One example would be if a user selects all items and unselects a subset, we could send a value representing the "Select all" alongside the subset of unselected items.

**Review by michael-s-molina — COMMENTED** [review]
Thanks for the PR @cccs-RyanK. I left some first-pass comments.

**Comment by EugeneTorap** [comment]
@cccs-RyanK any news about this PR?

File: superset-frontend/src/SqlLab/components/SqlEditorLeftBar/SqlEditorLeftBar.test.jsx
Comment (cccs-RyanK) [inline superset-frontend/src/SqlLab/components/SqlEditorLeftBar/SqlEditorLeftBar.test.jsx:97]: This test adds one item into a 'multi' select component and then selects it. With the new behaviour, that counts as selecting all which renders a custom tag with the total number of items selected. This test would fail because it expected the single tag to be rendered in the select component when that is no longer the expected behaviour, so I changed the test. 

Another option is to make it so that the select all functionality is disabled in the case where there is only one item in the options. Please let me know your thoughts

**Comment by EugeneTorap** [comment]
Hi @cccs-RyanK. Thank you for this PR!
Can we try to use 'Select All' in columns, table viz? Or it'll be next PR?

**Review by michael-s-molina — COMMENTED** [review]
Thanks for addressing the first-pass review comments @cccs-RyanK. Here are the second review comments:

When "Select all" is checked, we shouldn't show the +N tag.

Screen Shot 2022-12-12 at 12 09 38 PM

We're showing "Select All (12)" but there are 11 items. We shouldn't count the "Select All" option:

Screen Shot 2022-12-12 at 12 07 26 PM

**Comment by michael-s-molina** [comment]
When `allowNewOptions` is enabled, if you add a new option and remove it, "Select All" will be unchecked.


https://user-images.githubusercontent.com/70410625/207110358-091b7ad9-591b-44a5-a0ba-0891756b56ba.mov

File: superset-frontend/src/SqlLab/components/SqlEditorLeftBar/SqlEditorLeftBar.test.jsx
Comment (geido) [reply superset-frontend/src/SqlLab/components/SqlEditorLeftBar/SqlEditorLeftBar.test.jsx:97]: I think I like the idea of disabling the "select all" when only 1 option is available.

File: superset-frontend/src/components/Select/Select.tsx
Comment (geido) [inline superset-frontend/src/components/Select/Select.tsx:344]: Can we avoid the use of `any` here?

Comment (geido) [inline superset-frontend/src/components/Select/Select.tsx:344]:
```suggestion
    const selectAllModeTagRenderer = (props: any) => {
```

File: superset-frontend/src/components/Select/Select.tsx
Comment (geido) [inline superset-frontend/src/components/Select/Select.tsx:431]:
```suggestion
          tagRender={selectAllMode ? selectAllModeTagRenderer : undefined}
```

**Review by geido — COMMENTED** [review]
This is great! Just a few nits from a look to the code

**Comment by geido** [comment]
Not sure if this was discussed before, but I disagree with the choice of showing the number of items selected when "Select all" is checked and the plain items with the +X when the "Select all" is unchecked. I tend to believe we should always be showing the plain items that fit and the +X which is the default behaviour independently of whether the "Select all" option is checked to avoid the sudden change shown in the video and for consistency reasons

https://user-images.githubusercontent.com/60598000/207359271-47b47db1-f853-4362-b2c4-7170152b7931.mp4

When clearing the "Select all" from the options, the dropdown shows up. I think it shouldn't as the user intends to clear the selection and not to choose other options necessarily. Additionally, if you look at the video you will see that the dropdown in not positioned vertically in the correct way for a moment

https://user-images.githubusercontent.com/60598000/207360266-810a2b6a-9c6f-4033-b3ce-dc10085b93ae.mp4

When `allowNewOptions` is enabled if you add a new option, such as "test" shown in the video and you select it an unselect it, it's gone from the list (as shown in the video). 

https://user-images.githubusercontent.com/60598000/207360792-50cf73b8-8fc4-4698-b6be-10b355983931.mp4

If you create multiple ones and uncheck "Select all", all of them will be gone

https://user-images.githubusercontent.com/60598000/207361806-5243f9b0-9699-4509-8a7f-7b01fa97b497.mp4

File: superset-frontend/src/components/Select/Select.tsx
Comment (cccs-RyanK) [reply superset-frontend/src/components/Select/Select.tsx:325]: added isSingleMode and labelInValue as dependencies, but i purposefully left out fullSelectOptions.length because this hook should only be called when the component props have been changed. fullSelectOptions can change when new options are added and in that case we don't want the select value to be reset (erasing the newly added options).

**Comment by cccs-RyanK** [comment]
> When `allowNewOptions` is enabled If you create multiple options and uncheck "Select all", all of the newly created options will be erased. This should not be the case
> Select.-.Interactive.Select.Storybook.3.mp4

Thanks for the feedback @geido ! Made the suggested changes from your first two points. As for the last point you made about the new options being erased when they are deselected, I do think that this behaviour is consistent with how the component currently works. For example, before the select-all changes if you add new values by themselves and then deselect them, they are erased from the options. My thoughts are that deselecting-all should then also remove all the new options. Please let me know if that makes sense to you.

**Comment by EugeneTorap** [comment]
Hi @cccs-RyanK! This PR is important for us. Any news when we can merge it?

**Comment by villebro** [comment]
FYI @cccs-RyanK @EugeneTorap this is on my priority todo list, and I'm hoping to give it a proper review+test pass this weekend.

File: superset-frontend/src/components/Select/Select.tsx
Comment (michael-s-molina) [reply superset-frontend/src/components/Select/Select.tsx:325]: Good point. This is indicative that the effect should be broken into two unrelated operations. Like this:

```
useEffect(() => {
  setSelectValue(value);
}, [value]);

useEffect(() => {
  // if all values are selected, add select all to value
  if (
    !isSingleMode &&
    ensureIsArray(value).length === fullSelectOptions.length &&
    fullSelectOptions.length > 0
  ) {
    setSelectValue(
      labelInValue
        ? ([...ensureIsArray(value), selectAllOption] as AntdLabeledValue[])
        : ([
            ...ensureIsArray(value),
            SELECT_ALL_VALUE,
          ] as AntdLabeledValue[]),
    );
  }
}, [value, isSingleMode, labelInValue, fullSelectOptions.length]);
```

**Comment by michael-s-molina** [comment]
> Thanks for the feedback @geido ! Made the suggested changes from your first two points. As for the last point you made about the new options being erased when they are deselected, I do think that this behaviour is consistent with how the component currently works. For example, before the select-all changes if you add new values by themselves and then deselect them, they are erased from the options. My thoughts are that deselecting-all should then also remove all the new options. Please let me know if that makes sense to you.

@cccs-RyanK @geido I think we should disassociate the uncheck and remove actions on new items. We could have an X icon to remove new items. I can think of a scenario where a user is adding filters to a chart and after seeing the results, the user can check/uncheck the added filters to test different combinations. Making the remove action more intentional will improve UX. Let's tackle this in a separate PR. For now, I'm fine with removing all items when Select All is unchecked.

**Comment by michael-s-molina** [comment]
When working in `oneLine` mode, it's counting the Select All option.


https://user-images.githubusercontent.com/70410625/211862341-565316ce-bffa-42bc-8a99-f072d5908782.mov

File: superset-frontend/src/components/Select/CustomTag.tsx
Comment (michael-s-molina) [inline superset-frontend/src/components/Select/CustomTag.tsx:76]: To get rid of the typescript error:
```suggestion
      <Tag onMouseDown={onPreventMouseDown} {...(props as object)}>
```

File: superset-frontend/src/components/Select/AsyncSelect.tsx
Comment (michael-s-molina) [inline superset-frontend/src/components/Select/AsyncSelect.tsx:530]: Previously, we only used a custom tag renderer if `oneLine` was true. Does `customTagRender` preserve the same behavior?

Comment (cccs-RyanK) [reply superset-frontend/src/components/Select/AsyncSelect.tsx:530]: yes `customTagRender` is the same  as `oneLineTagRender`. It has the same behaviour except it does not render the select all tag.

Comment (michael-s-molina) [reply superset-frontend/src/components/Select/AsyncSelect.tsx:530]: What I mean to ask is if `customTagRender` has the same behavior as `tagRender={undefined}`

Comment (cccs-RyanK) [reply superset-frontend/src/components/Select/AsyncSelect.tsx:520]: Oh I see. The only difference with `customTagRender` than the default is the inclusion of the `onMouseDown` handler that prevents the dropdown from opening when clicking the tag icons. This was added for the `oneLine` mode and it seemed useful for the regular use case, but i can add a check to not include it when the component is not in `oneLine` mode?

Comment (michael-s-molina) [reply superset-frontend/src/components/Select/AsyncSelect.tsx:520]: Interesting. I like the behavior. No change is needed then.

**Review by michael-s-molina — APPROVED** [review]
LGTM. Thank you for all the hard work and for addressing all the comments. This feature is a nice addition to the component!

**Review by rusackas — APPROVED** [review]
Not sure why @michael-s-molina 's review didn't light up the merge button. Adding another approval to see if it unblocks.

**Review by jinghua-qa — APPROVED** [review]

**Comment by john-bodley** [comment]
@cccs-RyanK and @michael-s-molina I was wondering whether this feature could be configurable, especially as it pertains to SQL Lab. For example at Airbnb we have schemas which house thousands of tables/views (see attachment) and if the user accidentally chooses the `Select All` option`  SQL Lab will implode.

It's unclear if this option really has any value in the context of SQL Lab, i.e., whether one would ever needs to see the schema for all the tables/views as opposed to selecting the subset to entities they're interested in.

Screenshot 2023-05-18 at 9 58 53 AM

**Comment by villebro** [comment]
I agree with @michael-s-molina , let's disable it in that particular context.

**Pull request merged**
